library/src/blas1/rocblas_scal.hpp

```diff
@@ -0,0 +1,111 @@
+/* ************************************************************************
```

Comment:
I thought we were merging the `.hpp` files into the `.cpp` files. It will be easier to maintain if the `*_kernel` and `*_device` functions are in the same `.cpp` file as the C and C++ interfaces (the C, `*_impl` and `*_template` functions).

If `*_template` needs to be called from another file, it does not need to be defined (just declared) in a header as long as the template is instantiated with the right arguments in the `.cpp` file.

Reply:
I would vote against merging the templates into .cpp if we want to promote reuse or customization.  If one wants to derive a new variant they would want to include an .hpp.

Reply:
Since the _template function is being called from `rocblas_scal_impl` (in rocblas_scal.cpp), `rocblas_scal_batched_impl` (in rocblas_scal_batched.cpp) and `rocblas_scal_strided_batched_impl` (in rocblas_scal_strided_batched.cpp) doesn't it make more sense to have one .hpp file which contains the shared code?

I think it would be easier to maintain by having a separate .hpp file with the shared functions, rather than putting them into one specific (non-batched) .cpp file, but perhaps I'm missing a disadvantage to this.

Reply:
> I would vote against merging the templates into .cpp if we want to promote reuse or customization. If one wants to derive a new variant they would want to include an .hpp.

**That's only true for class templates / containers / etc.**

Reuse can be done even if they are in the `.cpp` file, _for function templates_. 

You simply declare (not define) the function template in the `.hpp` file, and then define it in the `.cpp` file. You can explicitly instantiate it for whatever template arguments in needs in the `.cpp` file, but if they are the same template arguments as the common `extern "C"` wrappers' template arguments, you won't even need to explicitly instantiate them, as they will already be implicitly instantiated.

You then don't need to create the `.hpp` file until a function is needed outside of the `.cpp` file, in which case you create an `.hpp` file and declare, but don't define, the function template there.

For most rocBLAS routines, they aren't called from outside of their implementation, so keeping everything in the `.cpp` file leads to more compact code which is easier to maintain, since it's all in one place rather than scattered among several files.

Reply:
I wouldn't be opposed to declaring the template function in the `.hpp` file, even if it's not currently used outside of the `.cpp` file. I just don't like the definitions of the kernel functions being scattered among `.hpp` and `.cpp` files.

Reply:
For this PR, all three functions (`rocblas_scal`, `rocblas_scal_batched`, and `rocblas_scal_strided_batched`) all share a single kernel which is defined in the shared .hpp file. This follows how the reduction functions are defined as well - in a shared .h file in that case - along with other functions such as `rocblas_trtri` (rocblas_trtri_batched.hpp).
Moving forward, if we are to implement more batched/strided_batched functions, kernels will likely start to be shared more (between the batched/strided/non-batched versions of functions), moving .cpp implementation of kernels to .hpp files might be a good standard to start to address this.

Reply:
> Since the _template function is being called from `rocblas_scal_impl` (in rocblas_scal.cpp), `rocblas_scal_batched_impl` (in rocblas_scal_batched.cpp) and `rocblas_scal_strided_batched_impl` (in rocblas_scal_strided_batched.cpp) doesn't it make more sense to have one .hpp file which contains the shared code?
> 
> I think it would be easier to maintain by having a separate .hpp file with the shared functions, rather than putting them into one specific (non-batched) .cpp file, but perhaps I'm missing a disadvantage to this.

It could be separated out into a `rocblas_scal_template.cpp` file, but the only advantage over putting it into a `.hpp` file is that compile/link time might be reduced if the same template arguments are used more than once; the functionality remains the same as if it is kept in a `.hpp` file.

**If each `.cpp` file instantiates the template with _different_ template arguments, then each `.cpp` file will generate its own instantiation of the template, and there is no advantage to moving it from a `.hpp` file to a `.cpp` file.** If two or more `.cpp` files instantiate the template with the _same_ template arguments, then each `.cpp` file will have to reinstantiate and rebuild the template, while if the template is only instantiated in one `.cpp` file, it will only be instantiated and built once for each set of arguments, and then it is also available for other `.cpp` files to use.

As long as the calls of the template are already scattered among several `.cpp` and `.hpp` files, then unless it's easy and not too large to merge them into one `.cpp` file, then my concerns about scattered implementations will still remain, regardless of whether you put the `_template` function in a `.hpp` file or a `.cpp` file. 

Putting `_template` in a `.hpp` file:
- Advantages: Declaration = definition; possibly one less file
- Disadvantages: Rebuilds the template every time it is used; separates the implementation in the `.hpp` file away from the `.cpp` file(s) which use it

Putting `_template` in a `.cpp` file:
- Advantages: Builds the template only once for each set of template arguments, possibly reducing compile time; still allows it to be called from any `.cpp` file; allows putting all implementation variants of the kernel code into one `.cpp` file
- Disadvantages: Still requires a `.hpp` file declaring the template if it's used elsewhere, so possibly one extra file; code may be too large if merged into one `.cpp` file, causing the same file to be edited more often than if the files were separate

So, as long as there are multiple `.cpp` files using the same `.hpp` file, I'm neutral about keeping the `.hpp` file.